New rule: fix missing `git clone` when given a URL

MiguelGuthridge opened the pull request:
Often when I'm trying to `git clone` a project, I'll paste the URL into my terminal expecting it to also contain the `git clone` part. This rule should allow `thefuck` to detect the missing `git clone` and suggest it when someone pastes an SSH or HTTP/HTTP url that ends in `.git`.

A rule for correcting for double `git clone git clone [repo]` already exists. This rule addresses the opposite problem.

![image](https://user-images.githubusercontent.com/31365175/169957706-14e2c30d-5f84-4fb1-b59a-9b04e05db6eb.png)

I've also added some simple tests for it, which appear to pass.

This is my first PR with any meaningful change, so I'd appreciate some feedback on it. It still needs a bit of cleaning up, but I think it should be good to merge in once I remove a couple of random things

djh82 commented on thefuck/rules/git_clone_missing.py:
```diff
@@ -0,0 +1,26 @@
+from thefuck.specific.git import git_support
+
+output_contains = ': not found'
```
Fish returns `repo.git : command not found`

djh82 replied:
Base returns: `: No such file or directory`

djh82 replied:
So you might consider searching for `not found` or `no such file or directory`?

MiguelGuthridge replied:
Fixed with refactor

djh82 commented on thefuck/rules/git_clone_missing.py:
```diff
@@ -0,0 +1,26 @@
+# Note: could probably use better checking for URLs
+
+
+# @git_support
```
this commented out line can be removed

MiguelGuthridge replied:
Fixed :)

djh82 commented on tests/rules/test_git_clone_missing.py:
```diff
@@ -0,0 +1,53 @@
+
+https_url = 'https://github.com/nvbn/thefuck.git'
+http_url = 'http://github.com/nvbn/thefuck.git'
+ssh_url = 'git@github.com:nvbn/thefuck.git'
```
SSH cloning might also look like `ssh://git@github`, at which point I dont think your match rule will work as expected

MiguelGuthridge replied:
Fixed

djh82 commented on thefuck/rules/git_clone_missing.py:
```diff
@@ -0,0 +1,26 @@
+    )
+
+
+# @git_support
```
remove me

MiguelGuthridge commented on thefuck/rules/git_clone_missing.py:
```diff
@@ -0,0 +1,26 @@
+
+# @git_support
+def match(command):
+    if not output_contains in command.output:
```
Fixed with refactor

MiguelGuthridge commented on thefuck/rules/git_clone_missing.py:
```diff
@@ -0,0 +1,26 @@
+def match(command):
+    if not output_contains in command.output:
+        return False
+    return command.script.endswith('.git') and (
```
Good idea, implemented this. Turns out `urllib` is a bit fussy with SSH things so I had to do a bunch of yuckiness to get it working but it seems to work now

MiguelGuthridge commented:
@djh82 would appreciate a re-review. I think I've fixed up all the issues

scorphus reviewed (commented):
This rule should be output-independent. Why bother matching the output?

However, it reads quite specific to me, I don't think many users would benefit from it. A URL might be related to so many things, not only `git clone`.

The Fuck allows the user to have user-defined rules at their disposal: [Creating your own rules](https://github.com/nvbn/thefuck/#creating-your-own-rules).

Not every single rule has to be part of The Fuck.

MiguelGuthridge commented:
@scorphus thanks for your feedback! I believe that this rule would be worthwhile as I encountered it enough times to choose to spend my time on creating a rule for it. I tutor a computer science class where we teach git and have observed that I'm far from the only person to make this mistake. Since there's a rule for handling `git clone git clone`, which is useful when I type `git clone` then paste the URL, only for it to include its own `git clone` (Bitbucket does this), I'd say that this particular rule is reasonable enough, since it handles the opposite case.

Perhaps it could search for `'git'` within the URL as an additional measure to avoid false positives? This would still work for all most of the popular git servers I'm aware of, at least using SSH. Would that change make the addition be worthwhile? Can you think of any other methods I could use to reduce the number of false positives?

scorphus commented:
Thanks for sharing your testimonial. Let's get this merged.

Regarding matching the output, I thought about disregarding output altogether and caring only about `output.script` or `output.script_parts`.

Asserting `git` in the URL would blacklist bitbucket repos.

scorphus commented on thefuck/rules/git_clone_missing.py:
```diff
@@ -0,0 +1,42 @@
+
+Author: Miguel Guthridge
+'''
+from urllib import parse
```
We still support py2:

```suggestion
from six.moves.urllib import parse
```

scorphus commented on thefuck/rules/git_clone_missing.py:
```diff
@@ -0,0 +1,42 @@
+    output = command.output.lower()
+    script = command.script
+    # Check for a command not found error
+    if not any([search in output for search in output_contains]):
```
Let's not create a list just to iterate over it:

```suggestion
    if not any(search in output for search in output_contains):
```

scorphus commented on thefuck/rules/git_clone_missing.py:
```diff
@@ -0,0 +1,42 @@
+    if not any([search in output for search in output_contains]):
+        return False
+    # URLs can't have spaces
+    if ' ' in command.script:
```
Let's avoid iterating over that string once again:
```suggestion
    if len(command.script_parts) > 1:
```

scorphus commented on tests/rules/test_git_clone_missing.py:
```diff
@@ -0,0 +1,60 @@
+    [Command(c, e) for c, e in product(valid_urls, shell_errors)]
+)
+def test_match(cmd):
+    print(cmd)
```
```suggestion
```

scorphus commented on tests/rules/test_git_clone_missing.py:
```diff
@@ -0,0 +1,60 @@
+    [Command(c, shell_errors[0]) for c in invalid_urls]
+)
+def test_not_match(cmd):
+    print(cmd)
```
```suggestion
```

scorphus commented on tests/rules/test_git_clone_missing.py:
```diff
@@ -0,0 +1,60 @@
+    [Command(c, shell_errors[0]) for c in valid_urls]
+)
+def test_get_new_command(cmd):
+    print(cmd)
```
```suggestion
```

scorphus reviewed (changes requested):
Thanks for contributing, please consider my suggestions below.

Also, a new line in README is required.

MiguelGuthridge commented:
Thanks again for the feedback! I fixed up things as per your suggestion, and also removed the code that checked the output.

Double checking, since I'm not entirely sure, should it check the URL for the presence of `git` somewhere? The only case I can think of that wouldn't work would be cloning using HTTPS on Bitbucket where `.git` has been removed from the end of their copied URL. Their SSH username is `git`, and their copy clone button copies the `git clone` part anyway. It would be helpful for preventing false positives such as [YouTube videos](https://www.youtube.com/watch?v=dQw4w9WgXcQ).

scorphus commented on thefuck/rules/git_clone_missing.py:
```diff
@@ -0,0 +1,33 @@
+
+
+def match(command):
+    script = command.script
+    parts = command.script_parts
```
No need to care about Bash nor ZSH nor Fish, only `sh` and `cmd`/`pwsh`:
```suggestion
    # Check for a command not found error
    if which(command.script_parts[0]) or not (
        'not found' in command.output or 'is not recognized as' in command.output
    ):
        return False
    script = command.script
```
It's also marginally faster than using a generator with `any`.

Nit: declare variables only when necessary.

scorphus commented on thefuck/rules/git_clone_missing.py:
```diff
@@ -0,0 +1,33 @@
+    script = command.script
+    parts = command.script_parts
+    # We want it to be a URL by itself
+    if len(parts) > 1:
```
No need to create a new variable, that's readable enough:
```suggestion
    if len(command.script_parts) > 1:
```

scorphus commented on tests/rules/test_git_clone_missing.py:
```diff
@@ -0,0 +1,47 @@
+    assert not match(cmd)
+
+
+@pytest.mark.parametrize(
+    'cmd',
+    [Command(c, 'not found') for c in valid_urls]
+)
+def test_get_new_command(cmd):
+    assert get_new_command(cmd) == 'git clone ' + cmd.script
```
Would be better if the `Command` instance was created in the test itself rather than in the parameterization. Then `script` could be used instead of `cmd.script`:

```suggestion
@pytest.mark.parametrize('script', valid_urls)
def test_get_new_command(script):
    command = Command(script, shell_errors[0])
    new_command = 'git clone ' + script
    assert get_new_command(command) == new_command
```

scorphus commented on tests/rules/test_git_clone_missing.py:
```diff
@@ -0,0 +1,47 @@
+    assert match(cmd)
+
+
+@pytest.mark.parametrize(
+    'cmd',
+    [Command(c, 'not found') for c in invalid_urls]
+)
+def test_not_match(cmd):
+    assert not match(cmd)
```
Ditto.

scorphus reviewed (changes requested):
My previous review was partially written. I would conclude the thought that I initiated in the message before, but ended up forgetting about it. Apologies.

I'd complement by saying that checking the command output is actually desirable and I should've given it a bit more thought when commenting earlier. Doing so speeds things up for cases that don't match by a factor of 4, at least.

But I think it should be simplified. Please consider my suggestions below.

Also, please, could you bring back the tests cases that include the output? Thanks!

MiguelGuthridge commented:
Pretty sure I've got everything fixed up now! Let me know if there's anything else you'd like me to do!

MiguelGuthridge commented:
Just bumping this one again @scorphus 
Is there anything else you want me to do on it?

scorphus reviewed (approved):
Sorry for the delay, @MiguelGuthridge.

> Pretty sure I've got everything fixed up now! Let me know if there's anything else you'd like me to do!

You nailed it. :+1:

My bad: I've just realized that, for commands like `git@github.com:nvbn/thefuck.git` and `https://github.com/nvbn/thefuck.git`, [Macos's `sh`](# "GNU bash, version 3.2.57(1)-release-(x86_64-apple-darwin19)") errs with `/bin/sh: https://github.com/nvbn/thefuck.git: No such file or directory\n`.

I'll got ahead and make the necessary change and push it on top of yours. Let's get this merged.

scorphus commented:
@MiguelGuthridge, please let me know your thoughts about the last change I submitted.

MiguelGuthridge commented:
Looks pretty great to me!

scorphus reviewed (approved):
Thank you, @MiguelGuthridge! :shipit:

scorphus commented:
@MiguelGuthridge, thanks for the complete patch! Much appreciated! 🙌
